regparser/builder.py

```diff
@@ -32,6 +34,47 @@ def __init__(self, cfr_title, cfr_part, doc_number, checkpointer=None):
         self.doc_number = doc_number
         self.checkpointer = checkpointer or NullCheckpointer()
         self.writer = api_writer.Client()
+        self.notices_json = []
+        #self.notices = self.checkpointer.checkpoint('notice', lambda: fetch_notices(self.cfr_title, self.cfr_part, only_final=True))
+        self.notices = []
+        self.notice_doc_numbers = []
+        #self.eff_notices = group_by_eff_date(self.notices)
+        self.eff_notices = []
+
+    def fetch_notices_json(self):
+        self.notices_json = fetch_notice_json(self.cfr_title, self.cfr_part, only_final=True)
+        self.notice_doc_numbers = [notice_json['document_number']
+                                   for notice_json in self.notices_json]
+
+    def build_single_notice(self, notice_json, checkpoint=True):
+        logging.info('building notice {0} from {1}'.format(notice_json['document_number'], notice_json['full_text_xml_url']))
+        if checkpoint:
+            notice = self.checkpointer.checkpoint(
+                'notice-' + notice_json['document_number'],
+                lambda: build_notice(self.cfr_title, self.cfr_part, notice_json)
+            )
+        else:
+            notice = build_notice(self.cfr_title, self.cfr_part, notice_json)
+
+        return notice
+
+    def build_notice_from_doc_number(self, doc_number, checkpoint=True):
+        if doc_number in self.notice_doc_numbers:
+            notice_json = [notice for notice in self.notices_json
+                           if notice['document_number'] == doc_number][0]
+            notice = self.build_single_notice(notice_json, checkpoint)
+            self.notices.extend(notice)
+            self.eff_notices = group_by_eff_date(self.notices)
+            return notice
+
+    def build_notices(self, checkpoint=True):
+        for result in self.notices_json:
+            notice = self.build_single_notice(result, checkpoint)
+            self.notices.extend(notice)
+        modify_effective_dates(self.notices)
+        #   Only care about final
+        self.notices = [n for n in self.notices if 'effective_on' in n]
+        self.eff_notices = group_by_eff_date(self.notices)
```

Comment:
It looks like there's a lot more logic around fetching notices in the builder now and the public interface has been extended quite a bit. What would you think about separating this logic out into a separate module/object. Then we could remove all of the separate `self.*notice*` fields and replace them with a single interface.

On checking the `checkpoint` parameter, note that the checkpointer can be set to a Null object. That should allow you to keep your context in that field rather than adding a parameter to all the functions.